enhance firewalld modules and states with ipset, zone_absent, service_absent

remijouannet opened the pull request:
enhance firewalld modules and states with ipset, zone_absent, service_absent

### What does this PR do?

- Add service_absent and service_present, service become an alias to service_present
- Add zone_absent and zone_present, present become an alias to zone_present
- Add ipset_absent and ipset_present with necessary functions in firewalld module
- add option to some functions to skip check in order to speed up executions
- Add target option to zone_present

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the test documentation for details on how to implement tests
into Salt's test suite:
https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html -->
- [X] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
No

<!-- Please review Salt's Contributing Guide for best practices and guidance in
choosing the right branch:
https://docs.saltproject.io/en/master/topics/development/contributing.html -->

<!-- Additional guidance for pull requests can be found here:
https://docs.saltproject.io/en/master/topics/development/pull_requests.html -->

<!-- See GitHub's page on GPG signing for more information about signing commits
with GPG:
https://help.github.com/articles/signing-commits-using-gpg/ -->

remijouannet committed:
enhance firewalld modules and states with ipset, zone_absent,
service_absent

remijouannet committed:
add docstring zone_absent

remijouannet committed:
f

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+
+    .. code-block:: bash
+
+        salt '*' firewalld.get_ipsets
```
`firewalld.info_ipset`

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+
+def get_target(zone):
+    """
+    Get zone's target
```
Please document the arguments as well.

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+
+def set_target(zone, target, permanent=True):
+    """
+    Set zone's target
```
These arguments also need to be documented

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+
+def get_ipsets(permanent=True):
+    """
+    Print predefined ipsets
```
`permanent` needs a docs entry

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+
+def info_ipset(ipset):
+    """
+    Print ipset info
```
document the argument please

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+
+def new_ipset(ipset, ipset_type, family=None, options=None, restart=False):
+    """
+    Add a new ipset
```
document these arguments

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+
+def delete_ipset(ipset, permanent=True, restart=True):
+    """
+    Delete an existing ipset
```
add some docs for these arguments

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+
+def add_ipset_entry(ipset, entry):
+    """
+    Add an new entry to the specified ipset.
```
here as well

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+
+def remove_ipset_entry(ipset, entry):
+    """
+    Remove an entry from the specified ipset.
```
and here

twangboy reviewed (changes requested):
This will also need a changelog and some tests for the module and the state.

dwoz committed:
Add changelog and tests for firewalld ipset, zone_absent, service_absent, and target

Add changelog/67790.added.md and unit tests for:
- firewalld.zone_absent state (exists, absent, test mode)
- firewalld.service_absent state (exists, absent)
- firewalld.ipset_absent state (exists, absent)
- firewalld.ipset_present state (create+entries, idempotent)
- firewalld.zone_present with target option (change, idempotent)
- firewalld module: get_target, set_target, get_ipsets, new_ipset,
  delete_ipset, add_ipset_entry, remove_ipset_entry, info_ipset

remijouannet commented:
hello @dwoz, do you need some help for my MR ? i saw you've pushed some change

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+
+    for i in __firewall_cmd(cmd).splitlines():
+        if i.strip():
+            if re.match("^[a-z0-9]", i, re.I):
```
The Bug: ``re.match`` checks from the beginning of the string. Lines returned by ``firewall-cmd --info-ipset`` that look like ``type: hash:net`` or ``options: family=inet`` also match this regex because they start with alphanumeric characters (``t`` and ``o``).

The Impact: Every time the loop hits ``type: hash:net``, it will overwrite ``ipset_name`` to be ``"type: hash:net"``, completely breaking the dictionary structure and causing subsequent lines to write into an unexpected or non-existent key.

The Fix: It needs to explicitly isolate the header line (which doesn't contain a colon) or use a more precise regex block to identify the true ipset name declaration.

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+                            _ipset[ipset_name].update({id_: []})
+                        else:
+                            _ipset[ipset_name].update({id_: val.strip().split(" ")})
+                    elif id_ == "options":
+                        _ipset[ipset_name].update(
+                            {id_: dict(item.split("=") for item in val.split())}
+                        )
```
The Issue: If ``firewall-cmd`` outputs an unexpected option layout, or if an option flag doesn't explicitly contain an ``=`` symbol, ``item.split("=")`` will return a single-element list. Passing a single element to ``dict()`` will instantly throw a ``ValueError: dictionary update sequence element #0 has length 1; 2 is required`` and crash the execution module.

The Fix: Safely parse items by checking for the presence of ``=`` or split with a maxsplit parameter inside a comprehension wrapper.

twangboy commented on salt/modules/firewalld.py:
```diff
@@ -1041,3 +1041,194 @@ def remove_rich_rule(zone, rule, permanent=True):
+    return __parse_ipset(cmd)
+
+
+def new_ipset(ipset, ipset_type, family=None, options=None, restart=False):
```
The Issue: If a user calls ``new_ipset`` through a state or CLI without specifying a family, family defaults to ``None``. However, firewall-cmd requires an explicit network family (typically ``inet`` or ``inet6``) when defining certain ipset types like ``hash:net``.

The Fix: While letting ``None`` fall through is fine if the binary defaults cleanly under the hood, standard Salt patterns typically pass a safe default string (like ``inet``) or validate it explicitly to return a clean ``CommandExecutionError`` rather than letting the raw shell execution throw an opaque error downstream.

twangboy reviewed (changes requested):
Please address the earlier comments I made regarding documentation for parameters.
